File: platform/cli/src/format.ts
import chalk from 'chalk';
import type { DiffStatus, ChangeType, FeatureData, DiffData } from './workspace';

export function statusIcon(status: string): string {
  switch (status) {
    case 'success':
    case 'completed':
    case 'connected':
    case 'active':
      return chalk.green('\u2713');
    case 'failure':
    case 'failed':
    case 'error':
      return chalk.red('\u2717');
    case 'cancelled':
    case 'canceled':
    case 'skipped':
      return chalk.gray('\u2298');
    case 'queued':
    case 'pending':
    case 'waiting':
      return chalk.yellow('\u23f3');
    case 'in_progress':
    case 'running':
      return chalk.blue('\u25b6');
    default:
      return chalk.dim('\u2022');
  }
}

export function statusColor(status: string): chalk.Chalk {
  switch (status) {
    case 'success':
    case 'completed':
      return chalk.green;
    case 'failure':
    case 'failed':
    case 'error':
      return chalk.red;
    case 'cancelled':
    case 'canceled':
    case 'skipped':
      return chalk.gray;
    case 'queued':
    case 'pending':
    case 'waiting':
      return chalk.yellow;
    case 'in_progress':
    case 'running':
      return chalk.blue;
    default:
      return chalk.white;
  }
}

export function relativeTime(date: string | Date): string {
  const now = Date.now();
  const then = new Date(date).getTime();
  const diffMs = now - then;
  const diffSec = Math.floor(diffMs / 1000);
  const diffMin = Math.floor(diffSec / 60);
  const diffHour = Math.floor(diffMin / 60);
  const diffDay = Math.floor(diffHour / 24);

  if (diffSec < 5) return 'just now';
  if (diffSec < 60) return `${diffSec} seconds ago`;
  if (diffMin === 1) return '1 minute ago';
  if (diffMin < 60) return `${diffMin} minutes ago`;
  if (diffHour === 1) return '1 hour ago';
  if (diffHour < 24) return `${diffHour} hours ago`;
  if (diffDay === 1) return '1 day ago';
  return `${diffDay} days ago`;
}

export function duration(ms: number): string {
  if (ms < 1000) return `${ms}ms`;
  const sec = ms / 1000;
  if (sec < 60) return `${sec.toFixed(1)}s`;
  const min = Math.floor(sec / 60);
  const remainSec = sec % 60;
  if (min < 60) return `${min}m ${remainSec.toFixed(0)}s`;
  const hr = Math.floor(min / 60);
  const remainMin = min % 60;
  return `${hr}h ${remainMin}m`;
}

export function table(
  rows: string[][],
  headers?: string[],
): string {
  const allRows = headers ? [headers, ...rows] : rows;
  if (allRows.length === 0) return '';

  const colCount = Math.max(...allRows.map((r) => r.length));
  const colWidths: number[] = [];

  for (let c = 0; c < colCount; c++) {
    colWidths[c] = 0;
    for (const row of allRows) {
      const cell = row[c] || '';
      // Strip ANSI codes for width calculation
      const stripped = cell.replace(
        // eslint-disable-next-line no-control-regex
        /\u001b\[[0-9;]*m/g,
        '',
      );
      colWidths[c] = Math.max(colWidths[c]!, stripped.length);
    }
  }

  const lines: string[] = [];

  for (let i = 0; i < allRows.length; i++) {
    const row = allRows[i]!;
    const cells = row.map((cell, c) => {
      const stripped = cell.replace(
        // eslint-disable-next-line no-control-regex
        /\u001b\[[0-9;]*m/g,
        '',
      );
      const padding = (colWidths[c] || 0) - stripped.length;
      return cell + ' '.repeat(Math.max(0, padding));
    });
    lines.push(cells.join('  '));

    // Add separator after headers
    if (i === 0 && headers) {
      const sep = colWidths.map((w) => '\u2500'.repeat(w));
      lines.push(chalk.dim(sep.join('  ')));
    }
  }

  return lines.join('\n');
}

export function indent(text: string, level: number = 1): string {
  const prefix = '  '.repeat(level);
  return text
    .split('\n')
    .map((line) => prefix + line)
    .join('\n');
}

export function errorMessage(err: unknown): string {
  if (err instanceof Error) {
    if (err.name === 'ConnectionError') {
      return chalk.red(err.message);
    }
    return chalk.red(`Error: ${err.message}`);
  }
  return chalk.red(`Error: ${String(err)}`);
}

// ---------------------------------------------------------------------------
// Diff / Feature helpers
// ---------------------------------------------------------------------------

export function diffLabel(position: number): string {
  return `D${position}`;
}

export function diffStatusIcon(status: DiffStatus): string {
  switch (status) {
    case 'draft':     return '\u25cb';  // ○
    case 'submitted': return '\u25d1';  // ◑
    case 'approved':  return '\u2713';  // ✓
    case 'landed':    return '\u25cf';  // ●
    default:          return '\u2022';  // •
  }
}

export function featureTypePrefix(type: ChangeType): string {
  switch (type) {
    case 'feature': return 'feat';
    case 'fix':     return 'fix';
    case 'docs':    return 'docs';
    case 'chore':   return 'chore';
    default:        return 'feat';
  }
}

export function contextLine(feature: FeatureData | null, diff: DiffData | null): string {
  const parts: string[] = [];
  if (feature) parts.push(feature.branch || feature.slug);
  if (diff) parts.push(diffLabel(diff.position));
  return parts.join(' > ');
}

export function conflictMessage(file: string): string {
  return [
    `The file ${chalk.bold(file)} has conflicting changes.`,
    '',
    'To resolve:',
    `  1. Open ${chalk.cyan(file)} and look for conflict markers (<<<, ===, >>>)`,
    '  2. Keep the version you want and delete the markers',
    '  3. Save the file',
  ].join('\n');
}

export function nextStepHint(action: string): string {
  switch (action) {
    case 'diff':
      return chalk.dim('Run "devpod submit" to create a pull request, or "devpod diff" to add another change.');
    case 'sync':
      return chalk.dim('Your branch is up to date. Run "devpod diff" to continue working.');
    case 'edit':
      return chalk.dim('Run "devpod submit" to update the pull request with your edits.');
    case 'submit':
      return chalk.dim('Waiting for review. Run "devpod land" after approval to merge.');
    case 'land':
      return chalk.dim('Change landed. Run "devpod sync" to start fresh.');
    default:
      return '';
  }
}

File: platform/cli-go/internal/format/format.go
package format

import (
	"fmt"
	"math"
	"regexp"
	"strings"
	"time"

	"github.com/fatih/color"
)

// StatusIcon returns a colored icon for the given status string.
func StatusIcon(status string) string {
	switch status {
	case "success", "completed", "connected", "active":
		return color.GreenString("\u2713") // check mark
	case "failure", "failed", "error":
		return color.RedString("\u2717") // cross
	case "cancelled", "canceled", "skipped":
		return color.HiBlackString("\u2298") // circled dash
	case "queued", "pending", "waiting":
		return color.YellowString("\u23f3") // hourglass
	case "in_progress", "running":
		return color.BlueString("\u25b6") // play
	default:
		return color.HiBlackString("\u2022") // bullet
	}
}

// DiffLabel returns the display label for a diff position (e.g. "D1").
func DiffLabel(position int) string {
	return fmt.Sprintf("D%d", position)
}

// DiffStatusIcon returns an icon for a diff review status.
func DiffStatusIcon(status string) string {
	switch status {
	case "draft":
		return "\u25cb" // ○
	case "submitted":
		return "\u25d1" // ◑
	case "approved":
		return "\u2713" // ✓
	case "landed":
		return "\u25cf" // ●
	default:
		return "\u2022" // •
	}
}

// FeatureTypePrefix returns the conventional commit prefix for a change type.
func FeatureTypePrefix(t string) string {
	switch t {
	case "feature":
		return "feat"
	case "fix":
		return "fix"
	case "docs":
		return "docs"
	case "chore":
		return "chore"
	default:
		return "feat"
	}
}

// RelativeTime converts an ISO timestamp string to a human-readable relative time.
func RelativeTime(t string) string {
	parsed, err := time.Parse(time.RFC3339, t)
	if err != nil {
		// Try other common formats
		parsed, err = time.Parse("2006-01-02T15:04:05Z", t)
		if err != nil {
			parsed, err = time.Parse("2006-01-02 15:04:05 -0700", t)
			if err != nil {
				return t
			}
		}
	}

	diff := time.Since(parsed)
	seconds := int(math.Floor(diff.Seconds()))
	minutes := int(math.Floor(diff.Minutes()))
	hours := int(math.Floor(diff.Hours()))
	days := int(math.Floor(float64(hours) / 24))

	if seconds < 5 {
		return "just now"
	}
	if seconds < 60 {
		return fmt.Sprintf("%d seconds ago", seconds)
	}
	if minutes == 1 {
		return "1 minute ago"
	}
	if minutes < 60 {
		return fmt.Sprintf("%d minutes ago", minutes)
	}
	if hours == 1 {
		return "1 hour ago"
	}
	if hours < 24 {
		return fmt.Sprintf("%d hours ago", hours)
	}
	if days == 1 {
		return "1 day ago"
	}
	return fmt.Sprintf("%d days ago", days)
}

// Duration formats a millisecond count as a human-readable duration.
func Duration(ms int64) string {
	if ms < 1000 {
		return fmt.Sprintf("%dms", ms)
	}
	sec := float64(ms) / 1000.0
	if sec < 60 {
		return fmt.Sprintf("%.1fs", sec)
	}
	min := int(math.Floor(sec / 60))
	remainSec := math.Mod(sec, 60)
	if min < 60 {
		return fmt.Sprintf("%dm %.0fs", min, remainSec)
	}
	hr := min / 60
	remainMin := min % 60
	return fmt.Sprintf("%dh %dm", hr, remainMin)
}

// ansiStripRe matches ANSI escape codes for width calculation.
var ansiStripRe = regexp.MustCompile(`\x1b\[[0-9;]*m`)

// stripAnsi removes ANSI escape codes from a string.
func stripAnsi(s string) string {
	return ansiStripRe.ReplaceAllString(s, "")
}

// Table renders an aligned table. If headers is non-nil, a separator line
// is drawn after the first row.
func Table(headers []string, rows [][]string) string {
	var allRows [][]string
	if headers != nil {
		allRows = append(allRows, headers)
	}
	allRows = append(allRows, rows...)

	if len(allRows) == 0 {
		return ""
	}

	// Determine column count
	colCount := 0
	for _, row := range allRows {
		if len(row) > colCount {
			colCount = len(row)
		}
	}

	// Calculate column widths
	colWidths := make([]int, colCount)
	for c := 0; c < colCount; c++ {
		for _, row := range allRows {
			cell := ""
			if c < len(row) {
				cell = row[c]
			}
			stripped := stripAnsi(cell)
			if len(stripped) > colWidths[c] {
				colWidths[c] = len(stripped)
			}
		}
	}

	// Render rows
	var lines []string
	for i, row := range allRows {
		var cells []string
		for c := 0; c < colCount; c++ {
			cell := ""
			if c < len(row) {
				cell = row[c]
			}
			stripped := stripAnsi(cell)
			padding := colWidths[c] - len(stripped)
			if padding < 0 {
				padding = 0
			}
			cells = append(cells, cell+strings.Repeat(" ", padding))
		}
		lines = append(lines, strings.Join(cells, "  "))

		// Separator after headers
		if i == 0 && headers != nil {
			var sep []string
			for _, w := range colWidths {
				sep = append(sep, strings.Repeat("\u2500", w))
			}
			lines = append(lines, color.HiBlackString(strings.Join(sep, "  ")))
		}
	}

	return strings.Join(lines, "\n")
}

// Indent prepends each line of text with the given number of two-space indents.
func Indent(text string, spaces int) string {
	prefix := strings.Repeat("  ", spaces)
	var lines []string
	for _, line := range strings.Split(text, "\n") {
		lines = append(lines, prefix+line)
	}
	return strings.Join(lines, "\n")
}

// ConflictMessage returns a plain-language message about a conflict in a file.
func ConflictMessage(file string) string {
	bold := color.New(color.Bold).SprintFunc()
	cyan := color.CyanString(file)
	return strings.Join([]string{
		fmt.Sprintf("The file %s has conflicting changes.", bold(file)),
		"",
		"To resolve:",
		fmt.Sprintf("  1. Open %s and look for conflict markers (<<<, ===, >>>)", cyan),
		"  2. Keep the version you want and delete the markers",
		"  3. Save the file",
	}, "\n")
}

// NextStepHint returns a suggestion after a command completes.
func NextStepHint(action string) string {
	dim := color.HiBlackString
	switch action {
	case "diff":
		return dim("Run \"devpod submit\" to create a pull request, or \"devpod diff\" to add another change.")
	case "sync":
		return dim("Your branch is up to date. Run \"devpod diff\" to continue working.")
	case "edit":
		return dim("Run \"devpod submit\" to update the pull request with your edits.")
	case "submit":
		return dim("Waiting for review. Run \"devpod land\" after approval to merge.")
	case "land":
		return dim("Change landed. Run \"devpod sync\" to start fresh.")
	default:
		return ""
	}
}

// ErrorMsg returns a red error message.
func ErrorMsg(msg string) string {
	return color.RedString("Error: %s", msg)
}

// SuccessMsg returns a green success message.
func SuccessMsg(msg string) string {
	return color.GreenString(msg)
}

// WarnMsg returns a yellow warning message.
func WarnMsg(msg string) string {
	return color.YellowString(msg)
}

// DimText returns dim/gray text.
func DimText(msg string) string {
	return color.HiBlackString(msg)
}
